File: java/src/main/java/it/giannibombelli/wsc2026/payment/application/commands/RejectTransaction.java
package it.giannibombelli.wsc2026.payment.application.commands;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.Command;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;
import it.giannibombelli.wsc2026.payment.domain.payment.TransactionId;

public record RejectTransaction(PaymentId aggregateId, TransactionId transactionId, Description reason)
    implements Command<PaymentId> {
    public RejectTransaction {
        Require.requireArgument(aggregateId, "aggregateId");
        Require.requireArgument(transactionId, "transactionId");
        Require.requireArgument(reason, "reason");
    }
}

File: typescript/src/payment/application/commands/rejectTransaction.ts
// Command per rifiutare una transazione.

import type { Command } from '@/common/application/command.js';
import type { Description } from '@/common/domain/primitive/description.js';
import type { PaymentId } from '../../domain/payment/paymentId.js';
import type { TransactionId } from '../../domain/payment/transactionId.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export type RejectTransaction = Command<'RejectTransaction', PaymentId> & {
  readonly transactionId: TransactionId;
  readonly reason: Description;
};

export function rejectTransaction(
  aggregateId: PaymentId,
  transactionId: TransactionId,
  reason: Description
): RejectTransaction {
  requireArgument(aggregateId, 'aggregateId');
  requireArgument(transactionId, 'transactionId');
  requireArgument(reason, 'reason');
  return { kind: 'RejectTransaction', aggregateId, transactionId, reason };
}
